scipy/ndimage/tests/test_filters.py

```diff
@@ -11,7 +11,7 @@
 from pytest import raises as assert_raises
 
 from scipy import ndimage
-from scipy.ndimage.filters import _gaussian_kernel1d, rank_filter
+from scipy.ndimage.filters import _gaussian_kernel1d
```

Comment:
Ralf had [indicated that `scipy.ndimage.filters`](https://github.com/cupy/cupy/pull/5047#issuecomment-814914710) etc. are not intended to be public. All functions should be imported from `scipy.ndimage`. Requiring importing from the top level for to get the multimethod versions would seem in line with this.

In practice, I am sure there are many instances of downstream libraries not following this convention of only importing from `scipy.ndimage` itself (we had a few that I recently fixed in `scikit-image`). It probably doesn't help that `scikit-image` does not import very many things at the top level and requires explicitly importing from subpackages like `filters` or `morphology` instead!

Reply:
> Ralf had [indicated that `scipy.ndimage.filters`](https://github.com/cupy/cupy/pull/5047#issuecomment-814914710) etc. are not intended to be public. All functions should be imported from `scipy.ndimage`.

Just filed an issue for that: gh-14360

Reply:
I think when the same question came up for `fftpack`, someone linked to something like a "scipy cookbook" that imported from submodules. So, AFAIK these are assumed to be public API; intentional or not.

My opinion is either break `import scipy.ndimage.fiters` completely, or make it export the multimethods.

Reply:
If we make `scipy.ndimage.<module>` export the multimethods, we are actually giving more incentive and motivation for the new user to use these private modules. Instead, we can add a note in the docs as `Do not use this private namespace, instead import like so...` and deprecate these private modules which are public looking (without underscore ones) gradually.

Reply:
A documentation comment on a particular function isn't really enough IMO. Existing code will lie incorrect with nobody noticing and anyone who copies from existing code will perpetuate the discouraged usage. It needs to actually emit a warning or error.

>If we make `scipy.ndimage.<module>` export the multimethods, we are actually giving more incentive and motivation for the new user to use these private modules.

Having `scipy.ndimage.<module>.<func>`  behave differently than `scipy.ndimage.<func>` is a really bad idea IMO.

Reply:
I agree with emitting a warning in such a case. This is actually linked to gh-14360 anyway which proposes similar way and needs more discussion. It is probably not a good idea to suddenly break the use of private modules though.

> Having scipy.ndimage.<module>.<func> behave differently than scipy.ndimage.<func> is a really bad idea IMO.

I don't have any strong opinions here. @rgommers @grlee77 what do you think?

Reply:
gentle ping @rgommers.

Reply:
What do you think of the following approach?
   1.) `git move filters.py _filters.py` 
   2.) modify `_api.py` to import from `_filters` instead of `filters`
   3.) Make a new `filters.py` with a `__getattr__` similar to [Peter's suggestion here](https://github.com/scipy/scipy/pull/14419/files#r672433228), to call the multimethod from `_multimethods.py`. (The decision on adding a deprecation warning can be left to a separate PR)
   4.) repeat steps 1-3 for `morphology.py` and other subpackages

The suggestion above seems like it would require minimal refactoring in this PR, but still allow the multimethods to work if imported from subpackages.